### PR TITLE
Toggle Tab Visibility

### LINKED ISSUE #30314: Add toggleTabsVisibility to keybinding actions
- VSCode Version: Code 1.13.1 (379d2efb5539b09112c793d3d9a413017d736f89, 2017-06-14T18:13:05.928Z)
- OS Version: Darwin x64 16.1.0
- Extensions: the listing length exceeds browsers' URL characters limit
---

| Element | Toggled Setting | Keybinding Command | Contextually Located Keybinding*** |
|:-:|:-:|:-:|:-:|
| Tabs | `workbench.editor.showTabs`  | `workbench.action.toggleTabsVisibility` ** | `ctrl+cmd+w` |
| ActivityBar | `workbench.activityBar.visible` | `workbench.action.toggleActivityBarVisibility` | `ctrl+cmd+a` |
| Sidebar | * | `workbench.action.toggleSidebarVisibility` | `ctrl+cmd+s` |
| Statusbar | `workbench.statusBar.visible` | `workbench.action.toggleStatusBarVisibility` | `ctrl+cmd+x` |

\* As far I could tell, there is no *workbench.statusBar.visible*, only *workbench.sideBar.location*.

\*\* Proposed keybindings action 
* I wrote its prefix as such as the other toggleable areas are in `/workbench/browser/actions/` even though the settings is under `workbench.editor` and as I didn't find  ```editor/actions``` section or the sort. But I am unfamiliar with the conventions and structure of this project. ¯\\_(ツ)\_/¯.
  * Perhaps change `workbench.editor.showTabs` -> `workbench.tabBar.visible` to follow suite?

\*\*\* These are my own keybindings I use for the commands. Their positions on a keyboard (qwerty) make them easy to reach and physically represent what area of the editor they are toggling. 
* For windows users `cmd` -> `alt`

### PR DESCRIPTION
**Pull request by maxfurman**
Fixes #30314
Toggle the visibility of editor tabs, with keyboard shortcut ctrl-cmd-W

### CODE REVIEW

File: src/vs/workbench/browser/actions/toggleTabsVisibility.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+		@IConfigurationService private configurationService: IConfigurationService
+	) {
+		super(id, label);
+	}
+
+	public run(): TPromise<any> {
+		const visibility = this.configurationService.getValue<string>(ToggleTabsVisibilityAction.tabsVisibleKey);
+		const newVisibilityValue = !visibility;
+
+		return this.configurationService.updateValue(ToggleTabsVisibilityAction.tabsVisibleKey, newVisibilityValue, ConfigurationTarget.USER);
```
Comment (bpasero) [inline src/vs/workbench/browser/actions/toggleTabsVisibility.ts]: @sandy081 how is the state of things here, shouldn't we stay away from setting an explicit Configuration.TARGET and rather let the service decide where to update the value based on where it is defined?

@maxfurman the issue here is that this setting could be defined on a workspace level and then writing it back to user level would not cause any visual change.

**Review by bpasero — CHANGES_REQUESTED** [review]

File: src/vs/workbench/browser/actions/toggleTabsVisibility.ts (hunk 1 above)
Comment (sandy081) [reply src/vs/workbench/browser/actions/toggleTabsVisibility.ts]: @bpasero Yes, Do not use the target, if the action is intended to change the current effective value. Current state is that all existing actions are still using the target (debt item). Make sure new ones do not cause the debt.

Comment (bpasero) [reply src/vs/workbench/browser/actions/toggleTabsVisibility.ts]: @maxfurman can you make that change to not set the target?

Comment (maxfurman) [reply src/vs/workbench/browser/actions/toggleTabsVisibility.ts]: No problem. Simply removing the third argument to updateValue will do the trick, correct?

**Comment by bpasero** [comment]
Thanks 👍

**Comment by SidIcarus** [comment]
Thanks @maxfurman !
:D

**Pull request merged**